clients/src/main/java/org/apache/kafka/common/record/DefaultRecord.java

```diff
@@ -349,7 +351,16 @@ private static DefaultRecord readFrom(ByteBuffer buffer,
 
         if (numHeaders == 0)
             return new DefaultRecord(sizeInBytes, attributes, offset, timestamp, sequence, key, value, Record.EMPTY_HEADERS);
+        try {
```

Comment:
My point is that any of the calls to `readVarint` or even `buffer.position()` prior to reading the headers could also underflow. It seems like it would be useful to raise `InvalidRecordException` consistently in all cases (not just for headers).